src/libsyntax/parse/lexer/mod.rs

```diff
@@ -1455,8 +1456,9 @@ impl<'a> StringReader<'a> {
                         }
                     }
 
-                    self.fatal_span_verbose(start_with_quote, pos,
-                        String::from("character literal may only contain one codepoint")).raise();
+                    self.err_span_(start_with_quote, pos,
+                        "character literal may only contain one codepoint");
+                    self.bump();
```

Comment:
This `bump` also looks suspect. I'm trying a change locally, but I believe there's a high likelihood that removing this `self.bump()` will fix the problem with `let x = 'asdf //~^ ERROR '`.

Reply:
Ok, upon further review, just revert this one change and let's fatally fail on this case. The code is beyond meaninful recovery at this point. We'll keep the current behavior for unterminated `char` literals, but recover in all other cases.

Reply:
@JohnTitor I believe that with these changes we'll be as close as we can get quickly to the "ideal" output:

https://github.com/estebank/rust/commit/63ad0a4e8cce883b55db5557a8eba7bd80cc4e94

Reply:
I'm grateful for your support! Following your commit, I added some commits.